### ISSUE TITLE
sql: malloc of large anonymised statements takes significant CPU time

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#54459]
Take a case where a user has say a 20k item placeholder, e.g.

```
UPSERT INTO <table> (key, value) VALUES ($1, $2), ($3, $4)...($20000, $20001);
```

Running a lot of these in parallel results in a slowdown.

An investigation using a program such as this:
```
package main

import (
	"fmt"
	"strings"

	"github.com/jackc/pgx"
)

func main() {
	conf := pgx.ConnPoolConfig{
		ConnConfig: pgx.ConnConfig{
			Host:     "localhost",
			User:     "root",
			Database: "defaultdb",
			Port:     26257,
		},
	}
	db, err := pgx.NewConnPool(conf)
	if err != nil {
		panic(err)
	}
	const num = 1000
	strArr := make([]string, 0, num)
	intArr := make([]int, 0, num)
	intInfArr := make([]interface{}, 0, num)
	intPtrArr := make([]interface{}, 0, num)
	for i := 0; i < num; i++ {
		strArr = append(strArr, fmt.Sprintf("$%d::int", i+1))
		intArr = append(intArr, i+1)
		intInfArr = append(intInfArr, i+1)
		intPtrArr = append(intPtrArr, &intArr[i])
	}
	_, err = db.Prepare("my-query", fmt.Sprintf("select %s", strings.Join(strArr, ", ")))
	if err != nil {
		panic(err)
	}

	for {
		r := db.QueryRow("my-query", intInfArr...)
		err := r.Scan(intPtrArr...)
		if err != nil {
			panic(err)
		}
	}
}
```

And running a pprof on it shows that a large amount of time is spent on the `malloc` giving enough space to store the extremely long anonymized query. It's so big, the image cannot be uploaded on GitHub because the dimensions are too wide -- so inserting as a [private gdoc link](https://drive.google.com/file/d/1PIBhvgltPyfLTTL-NrlMso-02-kXfij1/view?usp=sharing).

Should we try reduce these kinds of anonymisations?

**Comment by knz** [comment]
@onno-vos-dev ^^ the answer to your 2nd question

**Comment by knz** [comment]
A solution to this issue could be to use a go `sync.Pool` for these allocations

**Comment by onno-vos-dev** [comment]
Thank you @otan for looking into this! 👍 And thanks @knz for the ping. Will be following this one as I'd prefer not having to go down the unnest path which quickly becomes nasty. I opened a support ticket earlier this evening (CEST time) and assume that lead to this so you have the comparison and analysis dumps from that. If not let me know and I can link the support ticket. 

Thank you both!

**Comment by otan** [comment]
This is the public version after investigation on your question :)

**Comment by otan** [comment]
hmm @onno-vos-dev i took a look at fixing this and realised your UPSERT case might be different. i'll pull up a new ticket tomorrow if i spin something new.

**Comment by onno-vos-dev** [comment]
@otan the UPSERT case is the only scenario where we have lots of parameters. For reading and deleting batches we use `ANY=($1)`. It should be fairly trivial to try to see if this bubbles up on SELECT and DELETE as well though by using `IN` instead but it's less efficient than `ANY` during our testing so we never even went with `IN ($1, $2...$N)`
